### PR TITLE
Add initial support for pointers in reverse mode

### LINKED ISSUE #195: Crash due to the usage of pointers in reverse mode
Minimum reproducible example:

```cpp
double f5(double a) {
  double *ptr;
  ptr = &a;
  return *ptr;
}
```
Another example is simply 

```cpp
double f5(double* p){
  return *p;
}
```

calling `clad::gradient` on the above results in the following error,

<details>
        <summary>stack trace</summary>

```text
DeclRefExpr for Decl not entered in LocalDeclMap?
UNREACHABLE executed at /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExpr.cpp:2619!
Stack dump:
0.	Program arguments: /home/grimmyshini/cern/obj/bin/clang-9 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all -disable-free -main-file-name temp.cpp -mrelocation-model static -mthread-model posix -mdisable-fp-elim -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -dwarf-column-info -debugger-tuning=gdb -resource-dir /home/grimmyshini/cern/obj/lib/clang/9.0.1 -I ../src/tools/clad/include/ -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/x86_64-linux-gnu/c++/9 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/x86_64-linux-gnu/c++/9 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/backward -internal-isystem /usr/local/include -internal-isystem /home/grimmyshini/cern/obj/lib/clang/9.0.1/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdeprecated-macro -fdebug-compilation-dir /home/grimmyshini/cern/obj -ferror-limit 19 -fmessage-length 0 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -fcolor-diagnostics -add-plugin clad -load /home/grimmyshini/cern/inst/lib/clad.so -faddrsig -o /tmp/temp-409fa2.o -x c++ ../src/tools/clad/demos/temp.cpp 
1.	<eof> parser at end of file
2.	LLVM IR generation of declaration 'f5_grad'
3.	Generating code for declaration 'f5_grad'
 #0 0x000055b1912b655f llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/grimmyshini/cern/src/lib/Support/Unix/Signals.inc:533:22
 #1 0x000055b1912b65f2 PrintStackTraceSignalHandler(void*) /home/grimmyshini/cern/src/lib/Support/Unix/Signals.inc:594:1
 #2 0x000055b1912b44e0 llvm::sys::RunSignalHandlers() /home/grimmyshini/cern/src/lib/Support/Signals.cpp:68:20
 #3 0x000055b1912b5f19 SignalHandler(int) /home/grimmyshini/cern/src/lib/Support/Unix/Signals.inc:385:1
 #4 0x00007faaf6d828a0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x128a0)
 #5 0x00007faaf59dcf47 raise /build/glibc-2ORdQG/glibc-2.27/signal/../sysdeps/unix/sysv/linux/raise.c:51:0
 #6 0x00007faaf59de8b1 abort /build/glibc-2ORdQG/glibc-2.27/stdlib/abort.c:81:0
 #7 0x000055b19123cfc7 bindingsErrorHandler(void*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /home/grimmyshini/cern/src/lib/Support/ErrorHandling.cpp:218:55
 #8 0x000055b1919da0f7 clang::CodeGen::CodeGenFunction::EmitDeclRefLValue(clang::DeclRefExpr const*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExpr.cpp:2624:21
 #9 0x000055b1919d2b25 clang::CodeGen::CodeGenFunction::EmitLValue(clang::Expr const*) (.localalias) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExpr.cpp:1274:50
#10 0x000055b1919d265a clang::CodeGen::CodeGenFunction::EmitCheckedLValue(clang::Expr const*, clang::CodeGen::CodeGenFunction::TypeCheckKind) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExpr.cpp:1213:22
#11 0x000055b191a1b3aa (anonymous namespace)::ScalarExprEmitter::EmitCheckedLValue(clang::Expr const*, clang::CodeGen::CodeGenFunction::TypeCheckKind) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:256:3
#12 0x000055b191a1b6bf (anonymous namespace)::ScalarExprEmitter::EmitLoadOfLValue(clang::Expr const*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:305:32
#13 0x000055b191a1c29a (anonymous namespace)::ScalarExprEmitter::VisitDeclRefExpr(clang::DeclRefExpr*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:508:30
#14 0x000055b191a336b0 clang::StmtVisitorBase<std::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::Visit(clang::Stmt*) /home/grimmyshini/cern/obj/tools/clang/include/clang/AST/StmtNodes.inc:919:1
#15 0x000055b191a1ba4d (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:424:57
#16 0x000055b191a24729 (anonymous namespace)::ScalarExprEmitter::VisitCastExpr(clang::CastExpr*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:2150:17
#17 0x000055b191a34dd9 clang::StmtVisitorBase<std::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::VisitImplicitCastExpr(clang::ImplicitCastExpr*) /home/grimmyshini/cern/obj/tools/clang/include/clang/AST/StmtNodes.inc:865:1
#18 0x000055b191a33608 clang::StmtVisitorBase<std::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::Visit(clang::Stmt*) /home/grimmyshini/cern/obj/tools/clang/include/clang/AST/StmtNodes.inc:865:1
#19 0x000055b191a1ba4d (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:424:57
#20 0x000055b191a30ead clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:4437:1
#21 0x000055b1919d20eb clang::CodeGen::CodeGenFunction::EmitPointerWithAlignment(clang::Expr const*, clang::CodeGen::LValueBaseInfo*, clang::CodeGen::TBAAAccessInfo*) (.localalias) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExpr.cpp:1142:10
#22 0x000055b1919da722 clang::CodeGen::CodeGenFunction::EmitUnaryOpLValue(clang::UnaryOperator const*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExpr.cpp:2687:54
#23 0x000055b1919d316a clang::CodeGen::CodeGenFunction::EmitLValue(clang::Expr const*) (.localalias) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExpr.cpp:1339:52
#24 0x000055b1919d265a clang::CodeGen::CodeGenFunction::EmitCheckedLValue(clang::Expr const*, clang::CodeGen::CodeGenFunction::TypeCheckKind) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExpr.cpp:1213:22
#25 0x000055b191a1b3aa (anonymous namespace)::ScalarExprEmitter::EmitCheckedLValue(clang::Expr const*, clang::CodeGen::CodeGenFunction::TypeCheckKind) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:256:3
#26 0x000055b191a1b6bf (anonymous namespace)::ScalarExprEmitter::EmitLoadOfLValue(clang::Expr const*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:305:32
#27 0x000055b191a1cc34 (anonymous namespace)::ScalarExprEmitter::VisitUnaryDeref(clang::UnaryOperator const*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:624:30
#28 0x000055b191a328e3 clang::StmtVisitorBase<std::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::Visit(clang::Stmt*) /home/grimmyshini/cern/src/tools/clang/include/clang/AST/StmtVisitor.h:91:26
#29 0x000055b191a1ba4d (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:424:57
#30 0x000055b191a24729 (anonymous namespace)::ScalarExprEmitter::VisitCastExpr(clang::CastExpr*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:2150:17
#31 0x000055b191a34dd9 clang::StmtVisitorBase<std::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::VisitImplicitCastExpr(clang::ImplicitCastExpr*) /home/grimmyshini/cern/obj/tools/clang/include/clang/AST/StmtNodes.inc:865:1
#32 0x000055b191a33608 clang::StmtVisitorBase<std::add_pointer, (anonymous namespace)::ScalarExprEmitter, llvm::Value*>::Visit(clang::Stmt*) /home/grimmyshini/cern/obj/tools/clang/include/clang/AST/StmtNodes.inc:865:1
#33 0x000055b191a1ba4d (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:424:57
#34 0x000055b191a30ead clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGExprScalar.cpp:4437:1
#35 0x000055b1919a5076 clang::CodeGen::CodeGenFunction::EmitScalarInit(clang::Expr const*, clang::ValueDecl const*, clang::CodeGen::LValue, bool) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGDecl.cpp:750:40
#36 0x000055b1919aaa1c clang::CodeGen::CodeGenFunction::EmitExprAsInit(clang::Expr const*, clang::ValueDecl const*, clang::CodeGen::LValue, bool) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGDecl.cpp:1886:19
#37 0x000055b1919aa6b2 clang::CodeGen::CodeGenFunction::EmitAutoVarInit(clang::CodeGen::CodeGenFunction::AutoVarEmission const&) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGDecl.cpp:1845:26
#38 0x000055b1919a7967 clang::CodeGen::CodeGenFunction::EmitAutoVarDecl(clang::VarDecl const&) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGDecl.cpp:1286:22
#39 0x000055b1919a27dc clang::CodeGen::CodeGenFunction::EmitVarDecl(clang::VarDecl const&) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGDecl.cpp:192:27
#40 0x000055b1919a24e4 clang::CodeGen::CodeGenFunction::EmitDecl(clang::Decl const&) (.localalias) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGDecl.cpp:137:47
#41 0x000055b19168b164 clang::CodeGen::CodeGenFunction::EmitDeclStmt(clang::DeclStmt const&) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGStmt.cpp:1132:3
#42 0x000055b19168745f clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGStmt.cpp:350:75
#43 0x000055b1916867e5 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGStmt.cpp:49:3
#44 0x000055b19168797c clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CGStmt.cpp:390:3
#45 0x000055b19170a753 clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CodeGenFunction.cpp:1033:36
#46 0x000055b19170b3c1 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CodeGenFunction.cpp:1198:21
#47 0x000055b191732ab1 clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CodeGenModule.cpp:4320:3
#48 0x000055b19172bdeb clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CodeGenModule.cpp:2759:47
#49 0x000055b19172adb8 clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CodeGenModule.cpp:2512:5
#50 0x000055b191736a7b clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (.localalias) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CodeGenModule.cpp:5128:37
#51 0x000055b1925774df (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/ModuleBuilder.cpp:161:7
#52 0x000055b19257093c clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CodeGenAction.cpp:172:7
#53 0x000055b191d41c9c clang::MultiplexConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /home/grimmyshini/cern/src/tools/clang/lib/Frontend/MultiplexConsumer.cpp:271:25
#54 0x00007faaf5561081 clad::plugin::CladPlugin::ProcessDiffRequest(clad::DiffRequest&) (.localalias) /home/grimmyshini/cern/src/tools/clad/tools/ClangPlugin.cpp:200:9
#55 0x00007faaf5560b9a clad::plugin::CladPlugin::HandleTopLevelDecl(clang::DeclGroupRef) /home/grimmyshini/cern/src/tools/clad/tools/ClangPlugin.cpp:128:7
#56 0x000055b191d41c9c clang::MultiplexConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /home/grimmyshini/cern/src/tools/clang/lib/Frontend/MultiplexConsumer.cpp:271:25
#57 0x000055b193bd546a clang::ParseAST(clang::Sema&, bool, bool) /home/grimmyshini/cern/src/tools/clang/lib/Parse/ParseAST.cpp:162:20
#58 0x000055b191d02587 clang::ASTFrontendAction::ExecuteAction() /home/grimmyshini/cern/src/tools/clang/lib/Frontend/FrontendAction.cpp:1041:11
#59 0x000055b19256f379 clang::CodeGenAction::ExecuteAction() /home/grimmyshini/cern/src/tools/clang/lib/CodeGen/CodeGenAction.cpp:1060:1
#60 0x000055b191d01eea clang::FrontendAction::Execute() /home/grimmyshini/cern/src/tools/clang/lib/Frontend/FrontendAction.cpp:938:38
#61 0x000055b191c9dc7e clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/grimmyshini/cern/src/tools/clang/lib/Frontend/CompilerInstance.cpp:944:42
#62 0x000055b191e61c19 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/grimmyshini/cern/src/tools/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:291:38
#63 0x000055b18fcf62bb cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/grimmyshini/cern/src/tools/clang/tools/driver/cc1_main.cpp:249:40
#64 0x000055b18fceb973 ExecuteCC1Tool(llvm::ArrayRef<char const*>, llvm::StringRef) /home/grimmyshini/cern/src/tools/clang/tools/driver/driver.cpp:309:64
#65 0x000055b18fcec051 main /home/grimmyshini/cern/src/tools/clang/tools/driver/driver.cpp:381:26
#66 0x00007faaf59bfb97 __libc_start_main /build/glibc-2ORdQG/glibc-2.27/csu/../csu/libc-start.c:344:0
#67 0x000055b18fcea15a _start (/home/grimmyshini/cern/obj/bin/clang-9+0x224415a)
clang-9: error: unable to execute command: Aborted (core dumped)
clang-9: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 9.0.1 (https://github.com/llvm-mirror/clang.git bed37d4b18ce6993fc2567e8de61f6b30fd0cb75) (https://github.com/llvm-mirror/llvm.git c2f5309b216cb02a0aae17353549d985b5b05635)
```
</details>

### LINKED ISSUE #197: Functions with pointer assigns lead to compilation error in reverse mode
Any assignment of pointers results in a compilation error. Minimum reproducable example:

```cpp
double f5(double* p) {
  double* another_ptr = p;
  return *another_ptr;
}
clad::gradient(f5);
```
results in 

```error: invalid operands to binary expression ('double *' and 'double *') ```

Ideally, these expressions should be ignored and warned against, they should still be compiled however. I suspect some issue with the ```_result``` type mismatch with the intermediate pointers, however, I am not sure. This issue is also closely related to #195 and solving that might also solve this one.

This and #195 both work towards adding array/reference differentiability support too and maybe solved as that feature is addressed. 

Another example to reproduce the same error is:

```cpp
double f5(double a) {
  double* ptr, *another_ptr;
  ptr =  another_ptr;
  return *ptr;
}
```
results in the same error when executed in reverse mode. It works as expected in the forward mode.

### PR DESCRIPTION
**Pull request by vaithak**
This commit adds support for pointer operation in reverse mode. The technique is to maintain a corresponding derivative pointer variable, which gets updated (and stored/restored) in the exact same way as the primal pointer variable in both forward and reverse passes.
Added a workaround (with a FIXME comment) in the UsefulToStoreGlobal method to essentially bypass TBR analysis results for pointer expr.

Fixes #195, Fixes #197

### CODE REVIEW

**Comment by vaithak** [comment]
@PetroZarytskyi can you provide some comments on what extra needs to be done to make a pointer update statement to be added in the final output when enable-tbr is on?

File: lib/Differentiator/VisitorBase.cpp [hunk 1]
```diff
@@ -781,4 +781,50 @@ namespace clad {
...
+  /// instead of:
+  ///
+  /// ```
+  /// _d_p + _d_i;
+  /// ```
+  ///
+  /// Therefore, effective derived expression of `i` is `i` instead of `_d_i`.
+  ///
+  /// This functions sets `derivedL` and `derivedR` arguments to effective
+  /// derived expressions.
```
Comment (vgvassilev) [inline lib/Differentiator/VisitorBase.cpp:814]: Can we use the opportunity to move the documentation in the header file?

File: test/FirstDerivative/UnsupportedOpsWarn.C [hunk 2]
```diff
@@ -34,23 +34,9 @@ int unOpWarn_0(int x){
 // CHECK-NEXT:   return 0;
 // CHECK-NEXT: }
 
-int unOpWarn_1(int x){
-    auto pnt = &x;  // expected-warning {{attempt to differentiate unsupported operator, ignored.}}
```
Comment (vgvassilev) [inline test/FirstDerivative/UnsupportedOpsWarn.C:38]: We have enabled this test practically, can we remove the warning and keep it?

**Review by vgvassilev — COMMENTED** [review]
Overall it looks good. Can you add more test cases to hit the codecov missing coverage report? Please add more information in the commit log to explain some of the basics. I suspect this commit can close a few bugs that were reported - can you go over them and see if they are fixed and enumerate them in the commit message with the `Fixes ` prefix?

File: test/FirstDerivative/UnsupportedOpsWarn.C (hunk 2 above)
Comment (vaithak) [reply test/FirstDerivative/UnsupportedOpsWarn.C:38]: I think this test file aims to test if the warnings for such operations are correctly outputted, now that we have 
The functionality tested here is already being tested in the test added in `Pointers.c`.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 3]
```diff
@@ -2081,35 +2096,38 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+          addToCurrentBlock(add_assign, direction::reverse);
         }
       }
-      // We should not output any warning on visiting boolean conditions
-      // FIXME: We should support boolean differentiation or ignore it
-      // completely
+      return {cloneE, derivedE};
+    } else {
       if (opCode != UO_LNot)
+        // We should not output any warning on visiting boolean conditions
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:2128]: We should warn if the boolean condition is on values dependent on the independent variable as that mathematically means discontinuity.

File: test/Gradient/Pointers.C [hunk 4]
```diff
@@ -20,6 +20,167 @@ double nonMemFn(double i) {
...
+// CHECK-NEXT:         *_d_p -= _r_d0;
+// CHECK-NEXT:     }
+// CHECK-NEXT:     {
+// CHECK-NEXT:         p = _t0;
+// CHECK-NEXT:         _d_p = _t1;
+// CHECK-NEXT:     }
+// CHECK-NEXT: }
+
+double arrayPointer(double* arr) {
+  double *p = arr;
```
Comment (parth-07) [inline test/Gradient/Pointers.C:61]: Can you please add a test with an expression such as `arr + i`?

I am concerned that this would not work because the derivative of an array (`arr`) would be of type `clad::array_ref<T>` and `clad::array_ref<T>` already has overloaded addition operator that means element-wise addition. We may need to transform `_d_arr + i` to `_d_arr.data() + i` when derivative type is of type `clad::array_ref`.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 5]
```diff
@@ -2306,6 +2327,14 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
       auto* Lblock = endBlock(direction::reverse);
       llvm::SmallVector<Expr*, 4> ExprsToStore;
       utils::GetInnermostReturnExpr(Ldiff.getExpr(), ExprsToStore);
+
+      // We need to store values of derivative pointer variables in forward pass
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp:2331]: This is inside the `else if` condition when `op` is an assignment operation. In that case, LHS is the one that is changing and one we need to store and restore.

File: lib/Differentiator/ReverseModeVisitor.cpp (hunk 3 above)
Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp]: updated 👍🏼

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 6]
```diff
@@ -2529,6 +2591,13 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
         if (initDiff.getExpr_dx())
           VDDerivedInit = initDiff.getExpr_dx();
       }
+      // if VD is a pointer type, then the initial value is set to the derived
+      // expression of the corresponding pointer type.
+      else if (isPointerType && VD->getInit()) {
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp:2598]: improved test cases and fixed pointers with const qualifiers (both ways).

File: test/Gradient/Pointers.C (hunk 4 above)
Comment (vaithak) [reply test/Gradient/Pointers.C]: Fixed this using `ptr_ref` method and also added tests for this.

File: test/Gradient/Pointers.C [hunk 7]
```diff
@@ -20,6 +18,263 @@ double nonMemFn(double i) {
...
+// CHECK-NEXT:    {
+// CHECK-NEXT:        p = _t0;
+// CHECK-NEXT:        _d_p = _t1;
+// CHECK-NEXT:    }
+// CHECK-NEXT: }
+
+double pointerParam(const double* arr, size_t n) {
+  double sum = 0;
+  for (size_t i=0; i < n; ++i) {
+    size_t* j = &i;
```
Comment (vaithak) [inline test/Gradient/Pointers.C:176]: @PetroZarytskyi, our workaround for allowing all pointer expressions fails here. As we need to track `i` too, even though it is not a pointer itself, it is referenced by a pointer. So, similar to reference variable types, we need to track the RHS too.

File: test/Gradient/Pointers.C (hunk 4 above)
Comment (parth-07) [reply test/Gradient/Pointers.C]: Awesome!

File: lib/Differentiator/ReverseModeVisitor.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (parth-07) [reply lib/Differentiator/ReverseModeVisitor.cpp:2333]: Oh. Thank you for the explanation.

File: include/clad/Differentiator/ArrayRef.h [hunk 8]
```diff
@@ -22,7 +22,7 @@ template <typename T> class array_ref {
 
 public:
   /// Delete default constructor
-  array_ref() = delete;
+  array_ref() = default;
```
Comment (parth-07) [inline include/clad/Differentiator/ArrayRef.h:25]: Why is this change required?

File: include/clad/Differentiator/ArrayRef.h [hunk 9]
```diff
@@ -58,6 +68,34 @@ template <typename T> class array_ref {
   /// Returns the reference to the underlying array
   CUDA_HOST_DEVICE T& operator*() { return *m_arr; }
 
+  // Increment and decrement operators - update the underlying pointer.
+  /// Prefix increment operator.
+  CUDA_HOST_DEVICE array_ref<T>& operator++() {
```
Comment (parth-07) [inline include/clad/Differentiator/ArrayRef.h:73]: I am not so sure about this change. Intuitively, `v + 1` and `v++` should give the same result. However, these two operations give completely different results for `array_ref`. 

I feel like we are making `clad::array_ref` all-in-one solution for different things. This pull request seems good to be merged, however, we should try to improve the design in the future.

File: include/clad/Differentiator/ArrayRef.h (hunk 8 above)
Comment (vaithak) [reply include/clad/Differentiator/ArrayRef.h:25]: If the primal contains an operation of the form `arr = arr + 1`, where arr is a pointer param, this will result in `d_arr = d_arr + 1` statement, where `d_arr` is of type `array_ref`.
Now, that we are updating this value, this means we need to store its old value and restore it in the reverse pass, thus, requiring a `clad::array_ref<T> _t0` initialization, which will require a default constructor.

File: include/clad/Differentiator/ArrayRef.h (hunk 9 above)
Comment (vaithak) [reply include/clad/Differentiator/ArrayRef.h:73]: Agreed. Instead of generating `++d_arr`, I will change the generated statement to `++d_arr.ptr_ref()` to keep it consistent.

File: include/clad/Differentiator/ArrayRef.h (hunk 8 above)
Comment (vgvassilev) [reply include/clad/Differentiator/ArrayRef.h:25]: Ideally we should not create references out of thin air.

File: include/clad/Differentiator/ArrayRef.h (hunk 9 above)
Comment (parth-07) [reply include/clad/Differentiator/ArrayRef.h:73]: That sounds great. I think it's fine if you make this change in a follow-up pull request.

Comment (vaithak) [reply include/clad/Differentiator/ArrayRef.h:73]: updated in this PR :)

File: include/clad/Differentiator/ArrayRef.h (hunk 8 above)
Comment (vaithak) [reply include/clad/Differentiator/ArrayRef.h:25]: Ideally, yes. But this is essentially the same restriction as we have for user-defined objects: https://github.com/vgvassilev/clad/issues/627.

File: test/Gradient/Pointers.C (hunk 7 above)
Comment (vaithak) [reply test/Gradient/Pointers.C:176]: Had a chat with Petro regarding this, we currently don't have a quick fix or workaround for this now. So, disabling tbr analysis for this test case seems to be the only option until we have good support of pointers in the analysis phase.

Comment (vgvassilev) [reply test/Gradient/Pointers.C:176]: Can we add a FIXME in the code and open an issue to keep track of it?

Comment (vaithak) [reply test/Gradient/Pointers.C:179]: added a FIXME add the top of this test file and also created an issue: https://github.com/vgvassilev/clad/issues/691

**Comment by vgvassilev** [comment]
Can you rebase and squash commits where necessary?

**Comment by vaithak** [comment]
> Can you rebase and squash commits where necessary?

updated 👍🏼

**Review by parth-07 — APPROVED** [review]
Looks good.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 10]
```diff
@@ -2827,6 +2920,11 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
       // If TBR analysis is off, assume E is useful to store.
       if (!enableTBR)
         return true;
+      // FIXME: currently, we allow all pointer operations to be stored.
+      // This is not correct, but we need to implement a more advanced analysis
+      // to determine which pointer operations are useful to store.
+      if (E->getType()->isPointerType())
+        return true;
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:2927]: Can we add a test for this branch to make code of happy?

Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp:2927]: This is essentially a workaround for some cases of pointers when enableTbr is used, but as mentioned in some comment above, this doesn't cover every case. Adding this to test is essentially adding check with enable-tbr flag for pointer cases, but it will pass for some test and fail for others. Hence, adding it to pointers.c test won't be possible.

Comment (vgvassilev) [reply lib/Differentiator/ReverseModeVisitor.cpp:2927]: Well in theory we can add a tbr mode test and mark it xfail.

Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp:2927]: Will try this out. I have very little idea about xfail.

**Comment by parth-07** [comment]
A few important points to note:

- This pull request adds `clad::array_ref<T>::ptr_ref` interface. We need a similar interface / functionality for `clad::array<T>` as well for the same reason. For example:

```cpp
double arr[2] = {...};
double *u = arr + 1; // arr derivative type is clad::array, therefore, _d_arr + 1 would be element-wise addition! 
```

- Pointer arithmetic in a function call argument causes a compilation crash.

```cpp
fn(arr + 0, ...)
```

Two separate causes are responsible for the crash:
  - restore-value analysis tries to restore the value of expression `arr + 0`. That obviously doesn't work because `arr + 0` is not an lvalue. @PetroZarytskyi We perhaps need a check to only restore the value of the function arguments that are  lvalue expressions.
  - Currently, clad tries to create a dummy variable `clad::array<double> _r0(_d_arr.ptr_ref() + 0)`. This doesn't work because `clad::array` cannot be initialized from just a pointer. It needs a size as well, And `_d_arr.ptr_ref() + 0` doesn't have size information. One other important thing to add is that this `clad::array<double> _r0(_d_arr.ptr_ref() + 0)` declaration is completely useless for automatic differentiation. It's only used by the error-estimation framework.

----------------------------------

I think we should handle these issues separately in subsequent pull requests. This pull request already contains too many significant changes.

File: lib/Differentiator/ReverseModeVisitor.cpp (hunk 10 above)
Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp:2927]: Added a new test for TBR analysis with pointers and marked it as xfail for now.
